### PR TITLE
Start refactoring a common UI pattern: recalculate some value only when

### PR DESCRIPTION
**Pull request by dabreegster**
a key changes.

It's quite verbose right now for some widgetry States to determine if
the mouse is over a particular type of object, check if that object has
changed since the last time, and calculate something based on it. This
introduces a new helper for expressing the pattern a little more
clearly.

@michaelkirk Actually, I think this approach is better. Less boilerplate, works for App/SimpleApp, no weird lifetime issues, more flexible to pass params to key/value functions as needed.

### CODE REVIEW

File: fifteen_min/src/viewer.rs [hunk 1]
```diff
@@ -65,68 +58,24 @@ impl State<SimpleApp> for Viewer {
...
-                    }
-                },
-                _ => None,
-            };
-
-            // Also update this to conveniently get an outline drawn
-            app.current_selection = self.hovering_on_bldg.as_ref().map(|h| ID::Building(h.id));
+            let isochrone = &self.isochrone;
+            self.hovering_on_bldg
+                .update(HoverOnBuilding::key(ctx, app), |key| {
```
Comment (michaelkirk) [inline fifteen_min/src/viewer.rs:63]: The semantics here remind me of `or_insert_with_key` for [HashMap](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with_key) and [BTreeMap](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_insert_with_key) - a name closer to that might be clearer - Something like:

```
self.hovering_on_bldg.get_or_insert_with_key(HoverOnBuilding::key(ctx, app), |key| {
    HoverOnBuilding::value(ctx, app, key, isochrone)
})
```

File: fifteen_min/src/viewer.rs [hunk 2]
```diff
@@ -277,3 +224,61 @@ fn build_panel(
...
+            let dashed_lines = polyline.dashed_lines(
+                Distance::meters(0.75 * scale_factor),
+                Distance::meters(1.0 * scale_factor),
+                Distance::meters(0.4 * scale_factor),
+            );
+            batch.extend(Color::BLACK, dashed_lines);
+        }
+
+        // Also update this to conveniently get an outline drawn
+        app.current_selection = Some(ID::Building(hover_id));
```
Comment (michaelkirk) [inline fifteen_min/src/viewer.rs:273]: Having `app.current_selection` mutated as a side effect of this method is a little sneaky IMO. I think it was less surprising near the `ctx.redo_mouseover()` call. 

Maybe you agree or maybe not, but if you do, now you can hoist it to the callsite, which is a nice testament to this new API vs #405 👍: 

```
.update(HoverOnBuilding::key(ctx, app), |key| {
    app.current_selection = Some(ID::Building(hover_id));
    HoverOnBuilding::new(ctx, app, key, isochrone)
});
```

File: map_gui/src/simple_app.rs [hunk 3]
```diff
@@ -287,3 +287,39 @@ impl State<SimpleApp> for SimpleWarper {
 
     fn draw(&self, _: &mut GfxCtx, _: &SimpleApp) {}
 }
+
+/// Store a cached key/value pair, only recalculating when the key changes.
+pub struct Cached<K, V> {
+    contents: Option<(K, V)>,
+}
+
+impl<K: PartialEq + Clone, V> Cached<K, V> {
```
Comment (michaelkirk) [inline map_gui/src/simple_app.rs:296]: Will we provide nicer compiler errors if we put these constraints on the struct as well?

**Review by michaelkirk — APPROVED** [review]
This is nice! I definitely prefer this over #405.

File: fifteen_min/src/viewer.rs (hunk 1 above)
Comment (michaelkirk) [reply fifteen_min/src/viewer.rs:63]: I guess `insert` is a weird thing for a non-collection.

I don't like `update` though... 😆 

maybe `get_or_set_with_key` ?

Comment (michaelkirk) [reply fifteen_min/src/viewer.rs:63]: To expound - I don't like `update` because if we're using the cached value, nothing is `update`d.

Comment (michaelkirk) [reply fifteen_min/src/viewer.rs:63]: I guess, thinking about it a bit more, it's not a getter. I think that's what I was failing to understand. I'm much less offended by `update` now. 😬

File: fifteen_min/src/viewer.rs (hunk 2 above)
Comment (dabreegster) [reply fifteen_min/src/viewer.rs]: Ah, good point, thanks! I successfully moved it into the closure, but then realized there's a bug -- we want to set current_selection to None, and the callback only happens for defined keys. So I just moved the setting out of the `update` call entirely.

File: map_gui/src/simple_app.rs (hunk 3 above)
Comment (dabreegster) [reply map_gui/src/simple_app.rs:296]: Done. I've been bitten by that before for sure. I guess it's unlikely somebody would pass around a `Cached<T>` in a different module than where they call `update`, but having the error show up immediately seems nicer.

File: fifteen_min/src/viewer.rs (hunk 1 above)
Comment (dabreegster) [reply fifteen_min/src/viewer.rs:63]: Right, it doesn't return anything. The naming could be `update_if_changed` or something, but I was hoping calling the type `Cached` makes this clear. Any other ideas, or is `update` OK?

**Comment by michaelkirk** [comment]
I’d assumed different semantics -“Update” as you have it is totally fine for these semantics.

File: fifteen_min/src/viewer.rs (hunk 2 above)
Comment (michaelkirk) [reply fifteen_min/src/viewer.rs]: oh right, good catch!

**Pull request merged**